File: apps/erpnext_fiscal_br/erpnext_fiscal_br/events/sales_invoice.py
"""
Eventos para Sales Invoice
"""

import frappe
from frappe import _


def on_submit(doc, method):
    """
    Executado quando uma Sales Invoice é submetida
    Pode ser usado para emissão automática de NFe
    """
    # Verifica se a empresa tem configuração fiscal
    from erpnext_fiscal_br.fiscal_br.doctype.configuracao_fiscal.configuracao_fiscal import ConfiguracaoFiscal
    
    config = ConfiguracaoFiscal.get_config_for_company(doc.company)
    
    if not config:
        return
    
    # Por enquanto, apenas marca como "Sem NF"
    # A emissão será feita manualmente pelo botão
    if not doc.get("status_fiscal"):
        frappe.db.set_value("Sales Invoice", doc.name, "status_fiscal", "Sem NF")


def on_cancel(doc, method):
    """
    Executado quando uma Sales Invoice é cancelada
    Verifica se há NFe vinculada
    """
    if doc.get("nota_fiscal"):
        nf = frappe.get_doc("Nota Fiscal", doc.nota_fiscal)
        
        if nf.status == "Autorizada":
            frappe.throw(
                _("Esta fatura possui uma NFe autorizada ({0}). "
                  "Cancele a NFe antes de cancelar a fatura.").format(nf.name)
            )


def validate(doc, method):
    """
    Validação adicional na Sales Invoice
    """
    pass


@frappe.whitelist()
def get_fiscal_status(sales_invoice):
    """
    Retorna o status fiscal de uma Sales Invoice
    
    Args:
        sales_invoice: Nome da Sales Invoice
    
    Returns:
        dict: Status fiscal
    """
    invoice = frappe.get_doc("Sales Invoice", sales_invoice)
    
    result = {
        "status_fiscal": invoice.get("status_fiscal", "Sem NF"),
        "nota_fiscal": invoice.get("nota_fiscal"),
        "chave_nfe": invoice.get("chave_nfe"),
        "numero_nfe": invoice.get("numero_nfe"),
        "serie_nfe": invoice.get("serie_nfe"),
        "protocolo": invoice.get("protocolo_autorizacao"),
    }
    
    if invoice.get("nota_fiscal"):
        nf = frappe.get_doc("Nota Fiscal", invoice.nota_fiscal)
        result["danfe"] = nf.danfe
        result["xml"] = nf.xml_autorizado
    
    return result


@frappe.whitelist()
def check_can_emit_nfe(sales_invoice):
    """
    Verifica se uma Sales Invoice pode emitir NFe
    
    Args:
        sales_invoice: Nome da Sales Invoice
    
    Returns:
        dict: Resultado da verificação
    """
    from erpnext_fiscal_br.services.validators import validate_sales_invoice_for_nfe
    
    return validate_sales_invoice_for_nfe(sales_invoice)

File: apps/erpnext_fiscal_br/erpnext_fiscal_br/public/js/sales_invoice.js
/**
 * Customizações para Sales Invoice
 * Adiciona botões de emissão de NFe/NFCe
 */

frappe.ui.form.on("Sales Invoice", {
    refresh: function(frm) {
        // Só mostra botões se a fatura estiver submetida
        if (frm.doc.docstatus !== 1) {
            return;
        }
        
        // Verifica se já tem NFe
        if (frm.doc.nota_fiscal) {
            // Mostra informações da NFe
            frm.add_custom_button(__("Ver NFe"), function() {
                frappe.set_route("Form", "Nota Fiscal", frm.doc.nota_fiscal);
            }, __("Nota Fiscal"));
            
            // Botão para DANFE
            if (frm.doc.status_fiscal === "Autorizada") {
                frm.add_custom_button(__("DANFE"), function() {
                    frappe.call({
                        method: "erpnext_fiscal_br.events.sales_invoice.get_fiscal_status",
                        args: { sales_invoice: frm.doc.name },
                        callback: function(r) {
                            if (r.message && r.message.danfe) {
                                erpnext_fiscal_br.abrir_danfe(r.message.danfe);
                            }
                        }
                    });
                }, __("Nota Fiscal"));
                
                // Botão para XML
                frm.add_custom_button(__("XML"), function() {
                    frappe.call({
                        method: "erpnext_fiscal_br.events.sales_invoice.get_fiscal_status",
                        args: { sales_invoice: frm.doc.name },
                        callback: function(r) {
                            if (r.message && r.message.xml) {
                                erpnext_fiscal_br.download_xml(r.message.xml, frm.doc.chave_nfe);
                            }
                        }
                    });
                }, __("Nota Fiscal"));
            }
            
            return;
        }
        
        // Adiciona botão de emissão de NFe
        frm.add_custom_button(__("Emitir NFe"), function() {
            // Verifica se pode emitir
            frappe.call({
                method: "erpnext_fiscal_br.events.sales_invoice.check_can_emit_nfe",
                args: { sales_invoice: frm.doc.name },
                callback: function(r) {
                    if (r.message) {
                        if (!r.message.valid) {
                            frappe.msgprint({
                                title: __("Não é possível emitir NFe"),
                                message: r.message.errors.join("<br>"),
                                indicator: "red"
                            });
                            return;
                        }
                        
                        // Mostra warnings se houver
                        if (r.message.warnings && r.message.warnings.length > 0) {
                            frappe.confirm(
                                __("Avisos:") + "<br>" + r.message.warnings.join("<br>") + 
                                "<br><br>" + __("Deseja continuar?"),
                                function() {
                                    emitir_nfe(frm);
                                }
                            );
                        } else {
                            emitir_nfe(frm);
                        }
                    }
                }
            });
        }, __("Nota Fiscal"));
        
        // Adiciona botão de emissão de NFCe
        frm.add_custom_button(__("Emitir NFCe"), function() {
            frappe.call({
                method: "erpnext_fiscal_br.events.sales_invoice.check_can_emit_nfe",
                args: { sales_invoice: frm.doc.name },
                callback: function(r) {
                    if (r.message) {
                        if (!r.message.valid) {
                            frappe.msgprint({
                                title: __("Não é possível emitir NFCe"),
                                message: r.message.errors.join("<br>"),
                                indicator: "red"
                            });
                            return;
                        }
                        
                        emitir_nfce(frm);
                    }
                }
            });
        }, __("Nota Fiscal"));
    },
    
    onload: function(frm) {
        // Atualiza indicador de status fiscal
        if (frm.doc.status_fiscal) {
            update_fiscal_indicator(frm);
        }
    }
});

function emitir_nfe(frm) {
    frappe.confirm(
        __("Confirma a emissão da NFe para esta fatura?"),
        function() {
            erpnext_fiscal_br.emitir_nfe(frm.doc.name, function(result) {
                frm.reload_doc();
            });
        }
    );
}

function emitir_nfce(frm) {
    frappe.confirm(
        __("Confirma a emissão da NFCe para esta fatura?"),
        function() {
            erpnext_fiscal_br.emitir_nfce(frm.doc.name, function(result) {
                frm.reload_doc();
            });
        }
    );
}

function update_fiscal_indicator(frm) {
    let indicator = "gray";
    let status = frm.doc.status_fiscal || "Sem NF";
    
    switch(status) {
        case "Autorizada":
            indicator = "green";
            break;
        case "Cancelada":
            indicator = "red";
            break;
        case "Rejeitada":
            indicator = "red";
            break;
        case "Pendente":
            indicator = "orange";
            break;
        default:
            indicator = "gray";
    }
    
    frm.page.set_indicator(status, indicator);
}
